src/v/datalake/coordinator/iceberg_file_committer.cc

```diff
@@ -94,6 +98,294 @@ get_iceberg_committed_offset(const iceberg::table_metadata& table) {
+                return file_committer::errc::failed;
+            }
+        } else {
+            pk.fields.push_back(std::nullopt);
+        }
+    }
+
+    return pk;
+}
+
+checked<iceberg::struct_value, file_committer::errc> build_partition_key_struct(
+  const model::topic& topic,
+  const iceberg::table_metadata& table,
+  const data_file& f) {
+    if (f.table_schema_id < 0) {
+        // File created by a legacy Redpanda version that only
+        // supported hourly partitioning, the partition key value is
+        // in the hour_deprecated field.
+        iceberg::struct_value pk;
+        pk.fields.emplace_back(iceberg::int_value(f.hour_deprecated));
+        return pk;
+    } else {
+        auto schema = table.get_schema(
+          iceberg::schema::id_t{f.table_schema_id});
+        if (!schema) {
+            vlog(
+              datalake_log.error,
+              "can't find schema {} in table for topic {}",
+              f.table_schema_id,
+              topic);
+            return file_committer::errc::failed;
+        }
+
+        return build_partition_key_struct(topic, table, f, *schema);
+    }
+}
+
+checked<iceberg::partition_key, file_committer::errc> build_partition_key(
+  const model::topic& topic,
+  const iceberg::table_metadata& table,
+  const data_file& f) {
+    auto pk_res = build_partition_key_struct(topic, table, f);
+    if (pk_res.has_error()) {
+        return pk_res.error();
+    }
+    return iceberg::partition_key{
+      std::make_unique<iceberg::struct_value>(std::move(pk_res.value()))};
+}
+
+template<typename Out>
+requires std::output_iterator<Out, iceberg::data_file>
+checked<std::nullopt_t, file_committer::errc> collect_icb_files(
+  const model::topic& topic,
+  const iceberg::table_metadata& table,
+  const iceberg::manifest_io& io,
+  const chunked_vector<data_file>& files,
+  Out&& icb_files) {
+    for (const auto& f : files) {
+        auto pk = build_partition_key(topic, table, f);
+        if (pk.has_error()) {
+            return pk.error();
+        }
+
+        // TODO: pass schema_id and pspec_id to merge_append_action
+        // (currently it assumes that the files were serialized with the
+        // current schema and a single partition spec).
+        *icb_files++ = {
+          .content_type = iceberg::data_file_content_type::data,
+          .file_path = io.to_uri(std::filesystem::path(f.remote_path)),
+          .file_format = iceberg::data_file_format::parquet,
+          .partition = std::move(pk.value()),
+          .record_count = f.row_count,
+          .file_size_bytes = f.file_size_bytes,
+        };
+    }
+
+    return std::nullopt;
+}
+
+ss::future<checked<std::nullopt_t, file_committer::errc>> commit_icb_files(
+  const iceberg::table_identifier& table_id,
+  iceberg::manifest_io& io,
+  iceberg::catalog& catalog,
+  const commit_offset_metadata commit_meta,
+  iceberg::table_metadata&& table,
+  chunked_vector<iceberg::data_file>&& icb_files) {
```

Comment:
pass by value? it doesn't seem like there is much of a performance advantage here?